components/serverless/api/v1beta1/function_types.go

```diff
@@ -0,0 +1,206 @@
+/*
+Copyright 2022.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1beta1
+
+import (
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// EDIT THIS FILE!  THIS IS SCAFFOLDING FOR YOU TO OWN!
+// NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
+
+type Runtime string
+
+const (
+	Python39 = "python39"
+	NodeJs14 = "nodejs14"
+)
+
+type Source struct {
+	// +optional
+	ArtifactSource *ArtifactSource `json:"artifactSource"`
+	//TODO: we should check if it is possible
+	// +optional
+	CRDSource *CRDSource `json:"crdSource"`
+	// +optional
+	GitRepositorySource *GitRepositorySource `json:"gitRepositorySource"`
+	// +optional
+	InlineSource *InlineSource `json:"inlineSource"`
+}
+
+type InlineSource struct {
+	Source       string `json:"source"`
+	Dependencies string `json:"dependencies"`
+}
+
+type GitRepositorySource struct {
+	Name      string `json:"name"`
+	Namespace string `json:"namespace"`
+}
+
+type CRDSource struct {
+	Api        string `json:"api"`
+	Version    string `json:"version"`
+	Kind       string `json:"kind"`
+	Name       string `json:"name"`
+	Namespaces string `json:"namespaces"`
+	Path       string `json:"path"`
+	BaseDir    string `json:"baseDir"`
+	// +optional
+	Credentials *Credentials `json:"credentials"`
+}
+
+type ArtifactSource struct {
+	//TODO: maybe it's worth do distinguish between local and external, as we know the internal URL
+	URL     string `json:"url"`
+	BaseDir string `json:"baseDir"`
+	// +optional
+	Credentials *Credentials `json:"credentials"`
+}
+
+type ConfigMapSelector struct {
+	Name      string `json:"name"`
+	Namespace string `json:"namespace"`
+}
+
+//TODO: To discuss auth options
+type Credentials struct {
+	// +optional
+	BasicAuth *BasicAuth `json:"basicAuth"`
+	// +optional
+	Oauth *Oauth `json:"oauth"`
+	// +optional
+	JWTToken *v1.SecretKeySelector `json:"token"`
+	// +optional
+	PersonalAccessToekn *v1.SecretKeySelector `json:"personalAccessToekn"`
+}
+
+//TODO: should we enforce using secret or give the ability to use configmaps or plain values
+type BasicAuth struct {
+	User     v1.SecretKeySelector `json:"user"`
+	Password v1.SecretKeySelector `json:"password"`
+}
+
+type Oauth struct {
+	OauthURL string               `json:"oatuhURL"`
+	User     v1.SecretKeySelector `json:"user"`
+	Password v1.SecretKeySelector `json:"password"`
+}
+
+type Template struct {
```

Comment:
What is this for?

Reply:
This is for defining annotations and labels on deployment of function.